Documentation for APICTL Common API Policy operations (#6103)
wso2/docs-apim: en/docs/install-and-setup/setup/api-controller/managing-common-api-policies/managing-common-api-policies.md

> # Managing Common API Policies

Reviewer: Let's add a small description under this to introduce "common API policies" and a summary of the ways we can manage them.

> ## Add Common API Policies in an environment

Reviewer: ```suggestion
## Add common API Policies in an environment
```

> ## Add common API Policies in an environment

Reviewer: Why would someone add a common API policy? Can we expand on the use case aspect here?

Author: I think this is already mentioned in the docs. Refer https://apim.docs.wso2.com/en/latest/design/api-policies/create-policy/

Reviewer: Let's add a link to this please.

Author: +1

Author: Fixed by [021c130](https://github.com/wso2/docs-apim/pull/6103/commits/021c13019ef5f9d516c36b1f4dd9088bf80d542e)

> You can add Common API Policies via the Publisher Portal.

Reviewer: ```suggestion
You can add common API Policies via the Publisher Portal.
```

> However, **WSO2 API Controller (apictl)** allows you to create API Policies without using the Publisher Portal. For more information on adding Common API Policies, see [Importing Common API Policies Via Dev First Approach]({{base_path}}/install-and-setup/setup/api-controller/managing-apis-api-products/importing-common-api-policies-via-dev-first-approach).

Reviewer: ```suggestion
However, **WSO2 API Controller (apictl)** allows you to create API Policies without using the Publisher Portal. For more information on adding common API Policies, see [Importing Common API Policies Via Dev First Approach]({{base_path}}/install-and-setup/setup/api-controller/managing-apis-api-products/importing-common-api-policies-via-dev-first-approach).
```

## Get Common API Policies in an environment

> Follow the instructions below to display a list of Common API Policies in an environment using apictl:

Reviewer: ```suggestion
Follow the instructions below to display a list of common API Policies in an environment using apictl:
```

> 1.  Make sure that the WSO2 API Manager (WSO2 API-M) 4.1.0 version is started and that the 4.1.0 version of apictl is set up.   

Reviewer: Let's avoid the version number. They will see this in the 4.1.0 docs so the version should be obvious to most users plus it will help when we cut newer branches in future.

     For more information, see [Download and Initialize the apictl]({{base_path}}/install-and-setup/setup/api-controller/getting-started-with-wso2-api-controller/#download-and-initialize-the-apictl).
2.  Log in to the WSO2 API-M in the environment by following the instructions in [Login to an Environment]({{base_path}}/install-and-setup/setup/api-controller/getting-started-with-wso2-api-controller/#login-to-an-environment).

> 3.  Run the corresponding apictl command below to get (list) Common API Policies in an environment.

Reviewer: ```suggestion
3.  Run the corresponding apictl command below to get (list) the common API Policies in an environment.
```

>     1. Get API Common Policies in an environment.

Reviewer: When we render this in the site, it will display as section 3.a. However, there is no section 3.b so a bullet point is better for this.

        -   **Command**
            ``` bash
            apictl get policies api -e <environment>
            ```
            ``` bash
            apictl get policies api --environment <environment>
            ```
            ``` bash
            apictl get policies api --environment <environment> --all 

>             ```

Reviewer: Is it better to have tabs here for these code blocks? They all do the same thing with some minor differences that we should explain.

Reviewer: Please add titles for each tab. Not sure if this will render, but here's an example of what I'm talking about:

Example: 
``` bash tab="Bash"
#!/bin/bash

echo "Hello world!"
```

``` c tab="C"
#include <stdio.h>

int main(void) {
  printf("Hello world!\n");
}
```

``` c++ tab="C++"
#include <iostream>
```

Author: Same comment as in https://github.com/wso2/docs-apim/pull/6103#discussion_r944140879

            !!! info
                **Flags:**  
                
                -   Required :  
                    `--environment` or `-e` : Environment to be searched  
                -   Optional :
                    `--all`: All Common API Policies available in the environment
                    `--limit` or `-l` : Maximum number of Common API Policies to return (Default 25)
                    `--format` : pretty-print environments using templates

            !!!note

>                 When executing the `apictl get policies api` command, using both the `--all` and `--limit` flags is not allowed.

Reviewer 2: ```suggestion
                When executing the `apictl get policies api` command, using both the `--all` and `--limit` flags at once is not allowed.
```

Author: Fixed by [b0088d6](https://github.com/wso2/docs-apim/pull/6103/commits/b0088d62782dbf2157f66977cc9dc6ad0df420c3)

            !!! example
                ```bash
                apictl get policies api -e dev 
                ```
                ```bash
                apictl get policies api --environment production --limit 15 
                ```    
                ```bash
                apictl get policies api --environment production
                ```  

>                 ```go

Reviewer: Why is this in go code?

                apictl get policies api --environment production --all 

>                 ```  

Reviewer: Same comment as above. Maybe tabs are better otherwise this is a sequence of code blocks and not that easy to read.

Reviewer 2: If we add tabs here, when the users are directly copying from here and paste on the terminal, the tab space will also get copied. I saw this in many places in our docs, even for deployment toml , XML configs etc [1]. It is really hard in user point of view to remove the tabs from the copied content and paste it somewhere. Shall we not include tabs here?
Also, IMO, this is not difficult to read when there are no tabs included. 
For example, see below screenshot. I can't see a difficulty in reading here.
![image](https://user-images.githubusercontent.com/25246848/184294225-64ddc115-037d-466e-b0d9-a27398cb2d6b.png)

[1] https://apim.docs.wso2.com/en/latest/reference/mediators/xslt-mediator/#configuration

Author: @samgnaniah WDYT? Do we have to use tabs?

Reviewer: Ok, let's exclude if so.

        -   **Response**

>             ```go

Reviewer: Similar question to above. Why "go"?

            ```
            
            !!! tip 
                When using the `apictl get policies api -e dev` command, `--all` optional flag can be used to 
                get all available Common API Policies.

            !!!note

>                 Output of the `get policies api` command can be formatted with Go Templates. For more information on formatting the get commands, see [Formatting the outputs of get commands]({{base_path}}/install-and-setup/setup/api-controller/advanced-topics/formatting-the-output-of-get-command).

Reviewer 2: Shall we mention that the --all flag and -l flag cannot be used at once?

Author: Fixed by [6465f5a](https://github.com/wso2/docs-apim/pull/6103/commits/6465f5a2f5f93d8ca2645700a95ab612cd0d25ed)

            !!!note
                Output of the `get policies api` command can be formatted with Go Templates. For more information on formatting the get commands, see [Formatting the outputs of get commands]({{base_path}}/install-and-setup/setup/api-controller/advanced-topics/formatting-the-output-of-get-command).
               

## Delete a Common API Policy in an environment

Follow the instructions below to delete a Common API Policy in an environment using apictl:

> 1.  Make sure that the WSO2 API-M 4.1.0 version is started and that the 4.1.0 version of apictl is set up.   

Reviewer: Same comment as above. Let's remove the version specific stuff.

    1. Delete a Common API Policy in an environment.

        -   **Command**
            ``` bash

>             apictl delete policy api -n <Common API Policy name> -e <environment>

Reviewer 2: Let's include the type flag in the doc too.

Reviewer 2: Use the format <Common API Policy-name>
Here, note that "Common API Policy" is the artifact name, so we can keep that as it is. "name" is an attribute of the common API policy, so we need to use a hyphen to append it to the main artifact.
Let's follow this format everywhere. Refer [1] for more information.

[1] https://apim.docs.wso2.com/en/4.1.0/install-and-setup/setup/api-controller/managing-apis-api-products/migrating-api-products-to-different-environments/

Author: You mean we have to add examples for `--name` as well right?

Reviewer 2: No changing the flag <Common API Policy name> to <Common API Policy-name>

Author: Fixed by [14ca1f8](https://github.com/wso2/docs-apim/pull/6103/commits/14ca1f88a1c77bb52d4234c692aef6534f17f45c)

            ```
            ``` bash
            apictl delete policy api --name <API name> --environment <environment> 
            ```

            !!! info
                **Flags:**  
                
                -   Required :  
                    `--environment` or `-e` : Environment from which the Common API Policy should be deleted  
                    `--name` or `-n` : Name of the Common API Policy to be deleted  

            !!!note

>                 In api controller v4.2.0, --version flag support is not provided. Instead, internally it considers the default version which is `v1` when finding the policy.

Reviewer 2: ```suggestion
                In apictl 4.1.0, --version flag support is not provided. Instead, internally it considers the default version which is `v1` when finding the policy.
```
Keep like this. We shouldn't use "api controller" anywhere. We already introduced the abbreviation "apictl" on the 6th line of this doc. Hence, we should use it directly.

Also, the versioning will be changed at once at the end during the release. Hence, keep it as 4.1.0.

Author: Fixed by [6465f5a](https://github.com/wso2/docs-apim/pull/6103/commits/6465f5a2f5f93d8ca2645700a95ab612cd0d25ed)

# Managing common API Policies

Following actions on common API Policies are allowed by apictl.

1. List common API Policies in an environment.
2. Export a common API Policy from an environment.
3. Import a common API Policy to an environment.

> 4. Delete a common API Policy fromn an environment.

Reviewer 3: typing mistake

Author: fixed via https://github.com/wso2/docs-apim/pull/6561/commits/54465ebaf1d2f4ae69c2fdfa50f1e1506b8c66b3

## Add common API Policies in an environment

> You can add common API Policies via the Publisher Portal. Check [Create API Policy via Publisher Portal]({{base_path}}/design/api-policies/create_policy).

Reviewer: ```suggestion
You can add common API Policies via the Publisher Portal. For more information, see [Create API Policy via Publisher Portal]({{base_path}}/design/api-policies/create_policy).
```
